com.unity.netcode.gameobjects/Tests/Editor/Messaging/MessageRegistrationTests.cs

```diff
@@ -179,5 +180,109 @@ public void WhenCreatingMessageSystem_BoundTypeMessageHandlersAreRegistered()
+                return listMessages;
+            }
+        }
+
+        [Test]
+        public void MessagesGetPrioritizedCorrectly()
+        {
+            var sender = new NopMessageSender();
+            var provider = new OrderingMessageProvider();
+            var messagingSystem = new MessagingSystem(sender, null, provider);
+
+            // the 3 priority messages should appear first, in lexicographic order
+            Assert.AreEqual(messagingSystem.MessageTypes[0], typeof(ConnectionApprovedMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[1], typeof(ConnectionRequestMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[2], typeof(OrderingMessage));
+
+            // the other should follow after
+            Assert.AreEqual(messagingSystem.MessageTypes[3], typeof(AAAEarlyLexicographicNetworkMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[4], typeof(zzzLateLexicographicNetworkMessage));
+
+            // there should not be any extras
+            Assert.AreEqual(messagingSystem.MessageHandlerCount, 5);
+
+            // reorder the zzz one to position 3
+            messagingSystem.ReorderMessage(3, XXHash.Hash32(typeof(zzzLateLexicographicNetworkMessage).FullName));
+
+            // the 3 priority messages should still appear first, in lexicographic order
+            Assert.AreEqual(messagingSystem.MessageTypes[0], typeof(ConnectionApprovedMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[1], typeof(ConnectionRequestMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[2], typeof(OrderingMessage));
+
+            // the other should follow after, but reordered
+            Assert.AreEqual(messagingSystem.MessageTypes[3], typeof(zzzLateLexicographicNetworkMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[4], typeof(AAAEarlyLexicographicNetworkMessage));
+
+            // there should still not be any extras
+            Assert.AreEqual(messagingSystem.MessageHandlerCount, 5);
+
+            // verify we get an exception when asking for an invalid position
+            try
+            {
+                messagingSystem.ReorderMessage(-1, XXHash.Hash32(typeof(zzzLateLexicographicNetworkMessage).FullName));
+                Assert.Fail();
+            }
+            catch (ArgumentException)
+            {
+            }
+
+            // reorder the zzz one to position 3, again, to check nothing bad happens
+            messagingSystem.ReorderMessage(3, XXHash.Hash32(typeof(zzzLateLexicographicNetworkMessage).FullName));
+
+            // the two non-priority should not have moved
+            Assert.AreEqual(messagingSystem.MessageTypes[3], typeof(zzzLateLexicographicNetworkMessage));
+            Assert.AreEqual(messagingSystem.MessageTypes[4], typeof(AAAEarlyLexicographicNetworkMessage));
+
+            // there should still not be any extras
+            Assert.AreEqual(messagingSystem.MessageHandlerCount, 5);
+        }
```

Comment:
Additional possible tests (would like to see these but it's up to you):

- ArgumentOutOfRange on ReorderMessage
- ReorderMessage for an unknown type (e.g. verify list gets an extra entry with null handler)
- Add or Verify there's a test for handling a message type that's not registered (null handler)

Reply:
I added the first 2. Third bullet point is hard to test. But code inspection tells us:

`MessagingSystem.cs`
```
                try
                {
                    handler.Invoke(reader, ref context, this);
                }
                catch (Exception e)
                {
                    Debug.LogException(e);
                }
```

Reply:
Yeah that third one was in reference to [this](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/2103#discussion_r938951264) comment about a special exception type for null handlers. The one you pasted has a comment implying the exception is due to issues in user code.

Reply:
```
        private class TestNoHandlerMessageProvider : IMessageProvider
        {
            public List<MessagingSystem.MessageWithHandler> GetMessages()
            {
                return new List<MessagingSystem.MessageWithHandler>
                {
                    new MessagingSystem.MessageWithHandler
                    {
                        MessageType = typeof(TestMessage),
                        Handler = null
                    }
                };
            }
        }

        [Test]
        public void WhenReceivingAMessageWithoutAHandler_ExceptionIsThrown()
        {
            m_MessagingSystem = new MessagingSystem(new NopMessageSender(), this, new TestNoHandlerMessageProvider());

            var messageHeader = new MessageHeader
            {
                MessageSize = (ushort)UnsafeUtility.SizeOf<TestMessage>(),
                MessageType = m_MessagingSystem.GetMessageType(typeof(TestMessage)),
            };
            var message = GetMessage();

            var writer = new FastBufferWriter(1300, Allocator.Temp);
            using (writer)
            {
                writer.TryBeginWrite(FastBufferWriter.GetWriteSize(message));
                writer.WriteValue(message);

                var reader = new FastBufferReader(writer, Allocator.Temp);
                using (reader)
                {
                    try
                    {
                        m_MessagingSystem.HandleMessage(messageHeader, reader, 0, 0, 0);
                    }
                    // TODO: Catch the specific exception once one is implemented ("HandlerNotRegisteredException"?)
                    catch (Exception)
                    {
                        return;
                    }

                    Assert.Fail();
                }
            }
        }
```

Reply:
Ok, got it. You mean to actually throw something, not throw-catch-log as would happen with existing block.

Reply:
hmm, yeah, i guess we have to swallow the exception, don't we? Otherwise we break the entire message pipeline and it'll be stuck on that message every loop. :(

I guess you're right we can't reasonably test for it - but I still think a separate, specific, log message is warranted as users won't have any idea where to even start debugging if they just get a null ref exception.